### PR TITLE
WIP: Firms update staging

### LINKED ISSUE #272: Add test cases
I had this recommendation a long time ago and I think this is the moment where we need to make some test cases for the bot.

### LINKED ISSUE #295: Add a template command for OPs
This is a recommendation of adding a new command - `!template link`, because the bot makes an automatic sticky post, realy valuable information, like OP meme's template can get lost. This new command will allow the OP to reply to the bot with a link to the template and bot will update it's sticky, by adding the link that OP has provided.

This means adding a new regex - `!template\s+(.+)`

If OP links to some kind of a porno or other nsfw/inappropriate material, there is little we can do to prevent that. The only thing we *can* do is make a blacklist of websites (porthub, xxvideos, and etc.). If an imgur link is attached to the submission, there is very very very little we can do to check the contents of the image.

I will start working on this when the right time comes.

### PR DESCRIPTION
**Pull request by mappum**
**EDIT:** This branch is now being treated as a staging branch for the firms update

----

*(Work in progress)*
Closes #272 
Closes #295

I'm setting up simple unit tests here to more easily finish the work on #271 without having to manually try each command on Reddit. This PR won't have 100% test coverage, but should save time by covering the core of the bot's behavior.

My plan is to mock out the `praw` comment API and `sqlachemy` session API enough to be able to run commands against the `CommentWorker`.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -48,3 +48,6 @@ services:
         image: mariadb:10.2.15
         env_file: .env
         restart: unless-stopped
+        command: ["mysqld", "--default-authentication-plugin=mysql_native_password"]
+        ports:
+            - 3306:3306
```
Comment (thecsw) [inline docker-compose.yml]: We don't want to expose our database! The credentials are open and database should be accessible only to the isolated network of containers

Comment (mappum) [reply docker-compose.yml]: Oops, didn't mean to commit that, was just exposing it locally for debugging. 😛

Comment (thecsw) [reply docker-compose.yml]: Okay!

File: test/mock_praw.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+class Redditor():
+    def __init__(self, name):
+        self.name = name
+
+    def __str__(self):
+        return self.name
+
+class Submission(str):
+    def __new__(cls, id, *args):
```
Comment (thecsw) [inline test/mock_praw.py]: l would recommend to change the `id` variable to something else as in python, `id` is a reserved keyword. It's really weird why they did that, idk. You can leave `id` is a var name in a method, so `self.id` or `comment.id` are okay to use but just using `id` alone (as a var) might be troublesome. You can rename it to `ID` or some C++ style `id_m`. Thanks

**Comment by thecsw** [comment]
Hey! So I tested the PR, looks fantastic! I fetched the branch and when I ran the shell file `./test.sh`, it gave me the following error:

```
======================================================================
ERROR: balance (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: balance
Traceback (most recent call last):
  File "/usr/lib/python3.7/unittest/loader.py", line 434, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/lib/python3.7/unittest/loader.py", line 375, in _get_module_from_name
    __import__(name)
  File "/home/thecsw/gits/memeinvestor_bot/test/balance.py", line 1, in <module>
    from test import Test
  File "/home/thecsw/gits/memeinvestor_bot/test/test.py", line 10, in <module>
    from comment_worker import CommentWorker
  File "src/comment_worker.py", line 12, in <module>
    import config
  File "src/config.py", line 19, in <module>
    USER_AGENT = os.environ['BOT_USER_AGENT']
  File "/usr/lib/python3.7/os.py", line 678, in __getitem__
    raise KeyError(key) from None
KeyError: 'BOT_USER_AGENT'
```

So, I tried to add all of the env vars to the shell script and export the values, just so `config.py` would ignore the errors if an env var is not found (you have to wove the export lines before the actual coverage test):

``` bash
sed -E '/[#]/d;/^$/d' .env.example | sed -E 's/.*/export &/g' >> test.sh
```

Then it ran! I got the logs, for some reason, some of them started with a dot:

```
INFO:root: -- testuser: !create
INFO:root: -- testuser: !invest 100
INFO:root: -- testuser: !invest 100
INFO:root: -- testuser: !active
.INFO:root: -- testuser: !create
INFO:root: -- testuser: !active
.INFO:root: -- testuser: !create
INFO:root: -- testuser: !balance
INFO:root: -- testuser: !balance
.INFO:root: -- testuser: !help
.INFO:root: -- testuser: !ignore
..INFO:root: -- testuser: !create
(43 lines were omitted for brevity)
INFO:root: -- autocreating
.INFO:root: -- testuser: !create
. (weird dot idk)
----------------------------------------------------------------------
Ran 26 tests in 9.319s
```

And then the final report came up:

```
OK
Name                              Stmts   Miss Branch BrPart  Cover
-------------------------------------------------------------------
src/api.py                          141    141     44      0     0%
src/calculator.py                    93     93     18      0     0%
src/comment_worker.py               210     89     80      7    52%
src/config.py                        23      0      0      0   100%
src/describe-tables.py                8      8      0      0     0%
src/formula-graphs-for-paper.py      29     29      4      0     0%
src/formula-test.py                  44     44      8      0     0%
src/formula-visualizer.py           134    134     44      0     0%
src/formula.py                       34     34      6      0     0%
src/kill_handler.py                   7      7      0      0     0%
src/main.py                          69     69     24      0     0%
src/message.py                      118     18     12      1    82%
src/models.py                        42      1      2      1    95%
src/stopwatch.py                     12     12      0      0     0%
src/submitter.py                     73     73     24      0     0%
src/tops1.py                         27     27      8      0     0%
src/utils.py                         45     23     14      4    44%
-------------------------------------------------------------------
TOTAL                              1109    802    288     13    25%
```

So, the question is, did you have that error with exporting the variables? I know that you have `source .env` and it should work but it throws an error on my machine. I execute the script from the top directiory of the repo, so `.env` is accessed via `./.env`

**UPDATE** AH! I got the error. I don't know why it errors out on different machines, on my machine it's triggered due to a whitespace in the `.env` file.

Example:
```
.env: line 110: by: command not found
```

.env: line 110:
```
BOT_USER_AGENT=thecswbot by u/thecsw
```

The whitespace makes the script go nuts, to fix this, simply quote the whole thing as a UNIX tradition:

```
BOT_USER_AGENT="thecswbot by u/thecsw"
```

Very cool, Matt!

**Comment by thecsw** [comment]
I tested the last firms testing and got the following error about too many diffs:

```
======================================================================
FAIL: test_active (active.TestActive)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/thecsw/gits/memeinvestor_bot/test/active.py", line 30, in test_active
    Investment('post2', 100, 100, time.time())
AssertionError: '\nYo[114 chars]:00:07 remaining)\n\n[#2](https://www.reddit.c[72 chars]g)\n' != '\nYo[114 chars]:00:09 remaining)\n\n[#2](https://www.reddit.c[72 chars]g)\n'
Diff is 815 characters long. Set self.maxDiff to None to see it.

----------------------------------------------------------------------
Ran 57 tests in 17.862s

FAILED (failures=1)
```

**Comment by mappum** [comment]
Oh interesting, just looks like the tests are running slowly so the timestamps are different. I'll make so we can pass in a fixed timestamp instead of using `time.time()` in the tests.

**Comment by mappum** [comment]
I merged the feature branches into here so we don't have to deal with the mess of branches later.

**Comment by thecsw** [comment]
Awesome!

**Review by thecsw — APPROVED** [review]
Approved, please do the honors! Don't forget to add your name to our list of contributors!

**Pull request merged**